### PR TITLE
Account for .buildinfo in repl when build-type: Configure

### LINKED ISSUE #9401: cabal repl (and nix build) do not respect <package>.buildinfo
**Describe the bug**
`cabal build` successfully builds [the cabal project](https://github.com/alexoundos/cabal-buildinfo-example), which uses `build-type: Configure` with `<package>.buildinfo` and CPP definition in it.
But `cabal repl` fails:
```console
$ cabal repl
Build profile: -w ghc-9.2.8 -O1
In order, the following will be built (use -v for more details):
 - cabal-buildinfo-example-0.1.0.0 (ephemeral targets)
Preprocessing library for cabal-buildinfo-example-0.1.0.0..
GHCi, version 9.2.8: https://www.haskell.org/ghc/  :? for help
[1 of 1] Compiling MyLib            ( src/MyLib.hs, interpreted )

src/MyLib.hs:6:14: error:
    Data constructor not in scope: MYDEF :: String
  |
6 | definition = MYDEF
  |              ^^^^^
Failed, no modules loaded.
```
Remarkably, `nix build` fails the same way!

**To Reproduce**

```
$ git clone https://github.com/alexoundos/cabal-buildinfo-example
$ cd cabal-buildinfo-example
$ nix develop
[devshell]$ cabal build
[devshell]$ cabal run
[devshell]$ cabal repl
```

**Expected behavior**
Generated `<project>.buildinfo` file must be respected in `cabal repl` (and `nix build`).

**System information**
 - Operating system: tested on Linux (NixOS 23.05)
 - `cabal`: v3.10.1.0
 - `ghc`: v9.2.8

**Additional context**

This issue was already reported: #4195, haskell/haskell-language-server#3780.

### PR DESCRIPTION
**Pull request by alt-romes**
In `autoconfUserHooks` we were not updating the `preRepl` hook to read additional build information from /package/@.buildinfo@.

Fixes #9401

### CODE REVIEW

**Review by geekosaur — APPROVED** [review]

**Comment by geekosaur** [comment]
You still need a second review. But if you want to block merging, mark it as a draft. (On the right sidebar)

**Comment by Mikolaj** [comment]
@alt-romes: FYI: https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#github-pull-request-conventions

**Review by andreabedini — COMMENTED** [review]
:+1:

How do we know if we have missed another one? Are these meant to be all the `pre` hooks? What about `preTest` and `preBench`?

**Comment by alt-romes** [comment]
@andreabedini  We're indeed missing `preTest` and `preBench`.

According to my judgement this is a historical oversight and we should read information from .buildinfo for preTest and preBench too (i.e. for all pre-hooks)

I'll fix the patch.

Unfortunately, I don't think there is a good systematic way of enforcing that all pre-hooks are overwritten, since the way these user hooks are built is incrementally from `emptyUserHooks` which still sets a lot of defaults. My hope is that by writing that we should read additional information from .buildinfo for all pre-hooks excepting for pre-conf it will become more obvious for future devs.

**Comment by alt-romes** [comment]
@andreabedini done, I've updated the patch to also read additional build information for preTest and preBench. Even if we cannot enforce that all pre-hooks that are not pre-conf are updated here, at least all existing pre-hooks are now accounted for.

**Review by Kleidukos — DISMISSED** [review]
This is missing a changelog entry, but apart from this I approve of the changes. :)

**Comment by alt-romes** [comment]
@Kleidukos I've added the changelog, I'm just missing your final approval.

**Review by Kleidukos — APPROVED** [review]
Fantastic, thank you!

**Comment by Kleidukos** [comment]
@mergifyIO backport 3.10

**Comment by andreasabel** [comment]
> Thanks! I'm trying to add a test, do not merge yet.

@alt-romes @Kleidukos What happened to the test?

**Pull request merged**
